File: compiler/nodes/patterns/or.go
package patterns

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type OrPatternNode struct {
	Patterns []database.Node
	Facts    *database.Facts

	matching database.Node
}

func (node *OrPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseOrPattern(parser *syntax.Parser) (*OrPatternNode, *syntax.Error) {
	span := parser.Spanned()

	many, err := syntax.ParseMany(parser, 2, ParsePatternElement, func(parser *syntax.Parser) (*struct{}, *syntax.Error) {
		_, err := parser.Token("OrOperator")
		if err != nil {
			return nil, err
		}

		parser.ConsumeLineBreaks()

		return &struct{}{}, nil
	})
	if err != nil {
		return nil, err
	}

	patterns := make([]database.Node, 0, len(many))
	for _, item := range many {
		patterns = append(patterns, item.Value)
	}

	return &OrPatternNode{
		Patterns: patterns,
		Facts:    database.NewFacts(span()),
	}, nil
}

func (node *OrPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	for _, pattern := range node.Patterns {
		visitor.Visit(pattern)
		visitor.Constraint(typecheck.GroupConstraint(node, pattern))
	}
}

func (node *OrPatternNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	c.WriteString(span, " && (false")
	for _, pattern := range node.Patterns {
		c.WriteString(span, " || (true")
		if err := c.Write(pattern); err != nil {
			return err
		}
		c.WriteString(span, ")")
	}
	c.WriteString(span, ")")

	return nil
}

func (node *OrPatternNode) EachTemporary(f func(database.Node)) {
	for _, pattern := range node.Patterns {
		EachTemporary(pattern, f)
	}
}

File: compiler/src/nodes/patterns/or.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Db, Fact, Node, NodeRef, Render},
    nodes::{InheritTemporaries, parse_pattern_element, visit_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug, Clone)]
pub struct InvalidOrPattern;

impl Fact for InvalidOrPattern {}

impl Render for InvalidOrPattern {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "is invalid `or` pattern")
    }
}

#[derive(Debug)]
pub struct OrPatternNode {
    pub patterns: Vec<NodeRef>,
}

impl Node for OrPatternNode {}

pub fn parse_or_pattern(parser: &mut Parser<'_>) -> Result<OrPatternNode, ParseError> {
    let patterns = parser
        .parse_many(2, parse_pattern_element, |parser| {
            parser.token(TokenKind::OrOperator)?;
            parser.consume_line_breaks();
            Ok(())
        })?
        .into_iter()
        .map(|(node, _)| node)
        .collect();

    Ok(OrPatternNode { patterns })
}

impl Visit for OrPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, None);

        if !visitor.current_match().allow_or {
            visitor.insert(node, InvalidOrPattern);
        }

        let prev_arm = visitor.current_match().arm.take();

        for pattern in &self.patterns {
            if prev_arm.is_some() {
                // HACK: This only works because `or` is only supported at the
                // top level, effectively making each pattern its own arm
                visitor.current_match().arm = Some(pattern.clone());
            } else {
                // Not checking for exhaustiveness here (e.g. inside an `is` expression)
            }

            visitor.visit(pattern);
            visitor.edge(pattern, node, "pattern");
            visitor.constraint(GroupConstraint::new(node.clone(), pattern.clone()));
        }

        visitor.current_match().arm = prev_arm;

        visitor.insert(node, InheritTemporaries(self.patterns.clone()));
    }
}

impl Codegen for OrPatternNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        ctx.write_string(" && (false");

        for pattern in &self.patterns {
            ctx.write_string(" || (true");
            ctx.write(pattern)?;
            ctx.write_string(")");
        }

        ctx.write_string(")");

        Ok(())
    }
}
